mathemagical-prototype.js

```diff
@@ -404,6 +413,76 @@ class Square {
   }
 }
 
+// TODO: refactor to use Vertex objects instead of p5 vectors
+class Line {
+  constructor(graphWindow, x1, y1, x2, y2) {
+    this.graphWindow = graphWindow;
+    this.verticesInGraph = [
+      createVector(x1, y1),
+      createVector(x2, y2)
+    ]
+    this.verticesInCanvas = this.getVerticiesInCanvas();
+    this.strokeWeight = graphWindow.getStrokeWeight();
+    this.isDraggable = false;
+  }
+
+  setIsDraggable(isDraggable) {
+    this.isDraggable = isDraggable;
+  }
+
+  getVerticiesInCanvas() {
+    return this.verticesInGraph.map(
+      v => createVector(this.graphWindow.X(v.x), this.graphWindow.Y(v.y))
+    );
+  }
+
+  computeGraphFromCanvas() {
+    return this.verticesInCanvas.map(
+      v => createVector(this.graphWindow.x(v.x), this.graphWindow.y(v.y))
+    )
+  }
+
+  getCenterInCanvas() {
+    const xCenter = (this.verticesInCanvas[0].x + this.verticesInCanvas[1].x)/2
+    const yCenter = (this.verticesInCanvas[0].y + this.verticesInCanvas[1].y)/2
+    return createVector(xCenter, yCenter)
+  }
+
+  getWidthInCanvas() {
+    return this.graphWindow.xScale * 5
+  }
+
+  getHeightInCanvas() {
+    return this.graphWindow.yScale * 5
+  }
+
+  takeInput(draggable) {
+    if (this.isDraggable) {
+      draggable.giveInput(this);
+    }
+    // TODO: each vertex needs to take input
+  }
+
+  translate(vector) {
+    for (let v of this.verticesInCanvas) {
+      v.add(vector)
+    }
+    
+    //reset vertices in graph window coordinates
+    this.verticesInGraph = this.computeGraphFromCanvas(this.verticesInCanvas);
+  }
+  
+  render() {
+    push();
+    const [start, end] = this.verticesInCanvas;
+    strokeWeight(this.strokeWeight);
+    line(start.x, start.y, end.x, end.y);
+    strokeWeight(3)
```

Comment:
Maybe the midpoint doesn't need to be displayed? Or if it does, would it take long to replace the hard-coded 3 with something else? It might be okay for the sake of the prototype.

Reply:
Midpoint display is to help indicate that you can drag it if your mouse is over that point.